core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala

```diff
@@ -191,6 +191,45 @@ class ReplicaManagerTest {
     }
   }
 
+  @Test
+  def testClearPurgatoryOnReceivingStopReplicaRequest() {
+    val props = TestUtils.createBrokerConfig(0, TestUtils.MockZkConnect)
+    props.put("log.dir", TestUtils.tempRelativeDir("data").getAbsolutePath)
+    val config = KafkaConfig.fromProps(props)
+    val mockLogMgr = TestUtils.createLogManager(config.logDirs.map(new File(_)), LogConfig(new Properties()))
+    val aliveBrokers = Seq(createBroker(0, "host0", 0), createBroker(1, "host1", 1))
+    val metadataCache = EasyMock.createMock(classOf[MetadataCache])
+    EasyMock.expect(metadataCache.getAliveBrokers).andReturn(aliveBrokers).anyTimes()
+    EasyMock.replay(metadataCache)
+    val rm = new ReplicaManager(config, metrics, time, kafkaZkClient, new MockScheduler(time), mockLogMgr,
+      new AtomicBoolean(false), QuotaFactory.instantiate(config, metrics, time, ""), new BrokerTopicStats,
+      metadataCache, new LogDirFailureChannel(config.logDirs.size))
+
+    try {
+      val brokerList = Seq[Integer](0, 1).asJava
+      // Make this replica the leader.
```

Comment:
Is it okay not to create partition/replica?

Reply:
I think so, as long as we populate the right `PartitionState` in `LeaderAndIsrRequest`, [`ReplicaManager.becomeLeaderOrFollower()`](https://github.com/linkedin/kafka/blob/45b1ec20b64a74d6938ca928052e5db7264a8640/core/src/main/scala/kafka/server/ReplicaManager.scala#L1067) will create a new `Partition` object if it is missing.